### DISCUSSION TITLE
LinearMouse now supports per-device settings!

### DISCUSSION THREAD

**Discussion by lujjjh** [linearmouse/linearmouse#163] [Announcements]
LinearMouse [v0.7.0-beta.1](https://github.com/linearmouse/linearmouse/releases/tag/v0.7.0-beta.1) is now available. This major release contains some notable changes.

## Known issues

This version is not fully compatible with macOS Catalina. It crashes when the Preferences window is opened. I will look into it soon.

## Brand new UI

To match the style of macOS Ventura, I designed a brand new UI for LinearMouse.

<p align="center"><img width="550" alt="image" src="https://user-images.githubusercontent.com/3000535/176170981-04b8764a-707d-4105-88f2-b9216bdf345f.png"></p>

The navigation buttons have been relocated to the sidebar, and an indicator of the current device has been added above.

## Per-device settings

You can choose which device to configure in the Preferences. By default, LinearMouse automatically chooses the active device (that is, the device you are using).

<p align="center"><img width="384" alt="image" src="https://user-images.githubusercontent.com/3000535/176171946-03fd8f3b-e5c9-4c41-87fc-6c908a2c39cf.png"></p>

To achieve this, I designed [a new configuration syntax](https://github.com/linearmouse/linearmouse/blob/v0.7.0-beta.1/Documentation/Configuration.md) for LinearMouse. If you are an advanced user, you can directly edit the JSON file to achieve more flexible and complex configurations.

## Configuration migration

You do not have to reconfigure your devices. LinearMouse will automatically migrate the configuration for you if LinearMouse is upgraded from previous versions.

Under the hood, it generates a configuration scheme for all mouse devices:

```json5
{
  "schemes" : [
    {
      "if" : {
        "device" : {
          "category" : "mouse"
        }
      },
      // ...omitted
  ]
}
```

You may remove the configuration file at `~/.config/linearmouse/linearmouse.json` if you want to clear the migration and reconfigure your devices.

## What's next?

Thanks to the new configuration syntax, per-app settings and more complex rules are also easy to implement, if the UI is not considered.

I will implement the functionalities first so that advanced users can configure it via JSON.

**Comment by jtylerhartley** [comment]
Hi @lujjjh! Thank you so much for Linear Mouse! I just downloaded the new release (version 7.0) and I am loving it so far! For my magic mouse I like the vertical and horizontal scrolling direction reversed but for one app it doesn't work quite right. Would it be possible to reverse the scrolling direction on an app by app basis in the future or is it something you can configure in the JSON file currently? Thanks!

**Reply by lujjjh** [reply]
Hey @jtylerhartley!

Glad you like it. I plan to support per-app settings in the next release. You may track it by subscribing #111.

**Reply by lujjjh** [reply]
LinearMouse now supports per-app settings: #189.

**Reply by jtylerhartley** [reply]
Wonderful, thank you! I will certainly try the feature out when it comes out of beta.

**Comment by TheRealPear** [comment]
Hello! Thank you for all your work on this app, it has improved my experience with using non-Apple mouses on macOS. I came across [this documentation](https://github.com/linearmouse/linearmouse/blob/main/Documentation/Configuration.md#app-matching) on configuring LinearMouse to specifically disable reverse scrolling while using an app. However, it seems like copying and pasting the exact setup causes VS Code to give me an error saying `Property scroll is not allowed.` Am I doing something wrong or is the documentation outdated?

<img width="305" alt="image" src="https://user-images.githubusercontent.com/20259871/178795824-4090883a-75c8-4f34-a185-79acf3ad02d1.png">

**Reply by TheRealPear** [reply]
I forgot to mention this in the original post, but I am using version 0.7.1 stable.

**Reply by lujjjh** [reply]
Thank you for pointing that out!

I've updated the documentation.

**Comment by framf0rt** [comment]
A very nice feature, I'm sure. Would be cool to allow a user to treat all external mice as one config. Kinda annoying having to config every new mouse I use.

**Reply by jborbik** [reply]
I double that. I hope this will be also considered.

**Reply by lujjjh** [reply]
Currently, this cannot be done through the GUI, as I haven't figured out how to design it in the GUI. However, you can manually modify the configuration file to achieve this. For more information, please refer to [Schemes](https://github.com/linearmouse/linearmouse/blob/main/Documentation/Configuration.md#schemes).

**Reply by jborbik** [reply]
That's awesome, thank you for quick reply!
